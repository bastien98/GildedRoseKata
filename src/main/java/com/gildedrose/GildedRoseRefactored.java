package com.gildedrose;

import java.util.List;

public class GildedRoseRefactored {
    List<Item> items;

    public GildedRoseRefactored(List<Item> items) {
        this.items = items;
    }

    public void updateQuality() {
        for (Item item : items) {
            if (item.name.equals("Aged Brie")) {
                if (item.quality < 50) {
                    item.quality = item.quality + 1;
                }
            }

            if(item.quality >0 && !item.name.equals("Sulfuras, Hand of Ragnaros") && !item.name.equals("Backstage passes to a TAFKAL80ETC concert") && !item.name.equals("Aged Brie") ) {
                item.quality = item.quality - 1;
            }

            if (item.name.equals("Backstage passes to a TAFKAL80ETC concert")) {
                updateBackstage(item);
            }

            if (!item.name.equals("Sulfuras, Hand of Ragnaros")) {
                item.sellIn = item.sellIn - 1;
            }

            if (item.sellIn < 0) {
                if (item.name.equals("Aged Brie")) {
                    if (item.quality < 50) {
                        item.quality = item.quality + 1;
                    }
                } else {
                    if (item.name.equals("Backstage passes to a TAFKAL80ETC concert")) {
                        item.quality = 0;
                    } else {
                        if (item.quality > 0) {
                            if (item.name.equals("Sulfuras, Hand of Ragnaros")) {
                                continue;
                            }
                            item.quality = item.quality - 1;
                        }
                    }
                }
            }
        }
    }

    private void updateBackstage(final Item item) {
        if (item.quality < 50) {
            item.quality = item.quality + 1;
        }

        if (item.sellIn < 11) {
            if (item.quality < 50) {
                item.quality = item.quality + 1;
            }
        }

        if (item.sellIn < 6) {
            if (item.quality < 50) {
                item.quality = item.quality + 1;
            }
        }
    }


}
