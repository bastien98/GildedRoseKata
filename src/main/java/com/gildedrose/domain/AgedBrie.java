package com.gildedrose.domain;


import static com.gildedrose.domain.ItemConfig.MAXIMUM_ITEM_QUALITY;

public class AgedBrie extends Item implements Updatable {

    public AgedBrie(Item item) {
        super(item.name, item.sellIn, item.quality);
    }

    public void update() {
        this.sellIn--;
        if (this.quality < MAXIMUM_ITEM_QUALITY) {
            this.quality++;
        }
        if (this.quality < MAXIMUM_ITEM_QUALITY && this.sellIn < 0) {
            this.quality++;
        }
    }
}
