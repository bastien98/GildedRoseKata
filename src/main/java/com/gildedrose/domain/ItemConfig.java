package com.gildedrose.domain;

public class ItemConfig {
    public static final int MINIMUM_ITEM_QUALITY = 0;
    public static final int MAXIMUM_ITEM_QUALITY = 50;
}
