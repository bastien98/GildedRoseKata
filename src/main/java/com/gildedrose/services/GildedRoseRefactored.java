package com.gildedrose.services;

import com.gildedrose.domain.Item;
import com.gildedrose.domain.ItemFactory;
import com.gildedrose.domain.SpecializedItem;

import java.util.ArrayList;
import java.util.List;

public class GildedRoseRefactored {
    private final List<SpecializedItem> items;

    public GildedRoseRefactored(List<Item> items) {
        this.items = new ArrayList<>();
        for (Item item : items) {
            this.items.add(ItemFactory.createSpecializedItem(item));
        }
    }

    public void updateQuality() {
        for (int i = 0; i < items.size(); i++) {
            items.get(i).updateItem();
        }
    }

    public List<SpecializedItem> getItems() {
        return items;
    }
}
