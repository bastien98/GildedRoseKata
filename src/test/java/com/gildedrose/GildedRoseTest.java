package com.gildedrose;

import com.gildedrose.domain.Item;
import com.gildedrose.services.GildedRose;
import com.gildedrose.services.GildedRoseRefactored;
import org.approvaltests.combinations.CombinationApprovals;
import org.junit.jupiter.api.Test;

import java.util.ArrayList;
import java.util.List;

import static org.junit.jupiter.api.Assertions.assertEquals;

class GildedRoseTest {

    @Test
    void orignalTest() {
        Item[] items = new Item[]{new Item("foo", 0, 0)};
        GildedRose app = new GildedRose(items);
        app.updateQuality();
        assertEquals("foo", app.items[0].name);
    }

    @Test
    void abc() {

        // Given
        String[] nameCombos = {
            "+5 Dexterity Vest",
            "Aged Brie",
            "Elixir of the Mongoose",
            "Sulfuras, Hand of Ragnaros",
            "Backstage passes to a TAFKAL80ETC concert"};

        Integer[] sellInCombos = {-1, 0, 5, 6, 7, 10, 11, 12};

        Integer[] qualityCombos = {-1, 0, 1, 49, 50};

        // When + Then
        CombinationApprovals.verifyAllCombinations(
            this::updateItemQuality,
            nameCombos,
            sellInCombos,
            qualityCombos
        );
    }

    private String updateItemQuality(String name, int sellIn, int quality) {
        List<Item> items = new ArrayList<>();
        items.add(new Item(name, sellIn, quality));
        GildedRoseRefactored gildedRoseRefactored = new GildedRoseRefactored(items);
        gildedRoseRefactored.updateQuality();
        return gildedRoseRefactored.items.get(0).toString();
    }


}
