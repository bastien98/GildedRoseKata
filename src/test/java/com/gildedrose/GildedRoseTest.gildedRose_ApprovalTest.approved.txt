[+5 Dexterity Vest, -1, 0] => +5 Dexterity Vest, -2, 0 
[+5 Dexterity Vest, -1, 1] => +5 Dexterity Vest, -2, 0 
[+5 Dexterity Vest, -1, 25] => +5 Dexterity Vest, -2, 23 
[+5 Dexterity Vest, -1, 49] => +5 Dexterity Vest, -2, 47 
[+5 Dexterity Vest, -1, 50] => +5 Dexterity Vest, -2, 48 
[+5 Dexterity Vest, 0, 0] => +5 Dexterity Vest, -1, 0 
[+5 Dexterity Vest, 0, 1] => +5 Dexterity Vest, -1, 0 
[+5 Dexterity Vest, 0, 25] => +5 Dexterity Vest, -1, 23 
[+5 Dexterity Vest, 0, 49] => +5 Dexterity Vest, -1, 47 
[+5 Dexterity Vest, 0, 50] => +5 Dexterity Vest, -1, 48 
[+5 Dexterity Vest, 1, 0] => +5 Dexterity Vest, 0, 0 
[+5 Dexterity Vest, 1, 1] => +5 Dexterity Vest, 0, 0 
[+5 Dexterity Vest, 1, 25] => +5 Dexterity Vest, 0, 24 
[+5 Dexterity Vest, 1, 49] => +5 Dexterity Vest, 0, 48 
[+5 Dexterity Vest, 1, 50] => +5 Dexterity Vest, 0, 49 
[+5 Dexterity Vest, 4, 0] => +5 Dexterity Vest, 3, 0 
[+5 Dexterity Vest, 4, 1] => +5 Dexterity Vest, 3, 0 
[+5 Dexterity Vest, 4, 25] => +5 Dexterity Vest, 3, 24 
[+5 Dexterity Vest, 4, 49] => +5 Dexterity Vest, 3, 48 
[+5 Dexterity Vest, 4, 50] => +5 Dexterity Vest, 3, 49 
[+5 Dexterity Vest, 5, 0] => +5 Dexterity Vest, 4, 0 
[+5 Dexterity Vest, 5, 1] => +5 Dexterity Vest, 4, 0 
[+5 Dexterity Vest, 5, 25] => +5 Dexterity Vest, 4, 24 
[+5 Dexterity Vest, 5, 49] => +5 Dexterity Vest, 4, 48 
[+5 Dexterity Vest, 5, 50] => +5 Dexterity Vest, 4, 49 
[+5 Dexterity Vest, 6, 0] => +5 Dexterity Vest, 5, 0 
[+5 Dexterity Vest, 6, 1] => +5 Dexterity Vest, 5, 0 
[+5 Dexterity Vest, 6, 25] => +5 Dexterity Vest, 5, 24 
[+5 Dexterity Vest, 6, 49] => +5 Dexterity Vest, 5, 48 
[+5 Dexterity Vest, 6, 50] => +5 Dexterity Vest, 5, 49 
[+5 Dexterity Vest, 7, 0] => +5 Dexterity Vest, 6, 0 
[+5 Dexterity Vest, 7, 1] => +5 Dexterity Vest, 6, 0 
[+5 Dexterity Vest, 7, 25] => +5 Dexterity Vest, 6, 24 
[+5 Dexterity Vest, 7, 49] => +5 Dexterity Vest, 6, 48 
[+5 Dexterity Vest, 7, 50] => +5 Dexterity Vest, 6, 49 
[+5 Dexterity Vest, 9, 0] => +5 Dexterity Vest, 8, 0 
[+5 Dexterity Vest, 9, 1] => +5 Dexterity Vest, 8, 0 
[+5 Dexterity Vest, 9, 25] => +5 Dexterity Vest, 8, 24 
[+5 Dexterity Vest, 9, 49] => +5 Dexterity Vest, 8, 48 
[+5 Dexterity Vest, 9, 50] => +5 Dexterity Vest, 8, 49 
[+5 Dexterity Vest, 10, 0] => +5 Dexterity Vest, 9, 0 
[+5 Dexterity Vest, 10, 1] => +5 Dexterity Vest, 9, 0 
[+5 Dexterity Vest, 10, 25] => +5 Dexterity Vest, 9, 24 
[+5 Dexterity Vest, 10, 49] => +5 Dexterity Vest, 9, 48 
[+5 Dexterity Vest, 10, 50] => +5 Dexterity Vest, 9, 49 
[+5 Dexterity Vest, 11, 0] => +5 Dexterity Vest, 10, 0 
[+5 Dexterity Vest, 11, 1] => +5 Dexterity Vest, 10, 0 
[+5 Dexterity Vest, 11, 25] => +5 Dexterity Vest, 10, 24 
[+5 Dexterity Vest, 11, 49] => +5 Dexterity Vest, 10, 48 
[+5 Dexterity Vest, 11, 50] => +5 Dexterity Vest, 10, 49 
[Aged Brie, -1, 0] => Aged Brie, -2, 2 
[Aged Brie, -1, 1] => Aged Brie, -2, 3 
[Aged Brie, -1, 25] => Aged Brie, -2, 27 
[Aged Brie, -1, 49] => Aged Brie, -2, 50 
[Aged Brie, -1, 50] => Aged Brie, -2, 50 
[Aged Brie, 0, 0] => Aged Brie, -1, 2 
[Aged Brie, 0, 1] => Aged Brie, -1, 3 
[Aged Brie, 0, 25] => Aged Brie, -1, 27 
[Aged Brie, 0, 49] => Aged Brie, -1, 50 
[Aged Brie, 0, 50] => Aged Brie, -1, 50 
[Aged Brie, 1, 0] => Aged Brie, 0, 1 
[Aged Brie, 1, 1] => Aged Brie, 0, 2 
[Aged Brie, 1, 25] => Aged Brie, 0, 26 
[Aged Brie, 1, 49] => Aged Brie, 0, 50 
[Aged Brie, 1, 50] => Aged Brie, 0, 50 
[Aged Brie, 4, 0] => Aged Brie, 3, 1 
[Aged Brie, 4, 1] => Aged Brie, 3, 2 
[Aged Brie, 4, 25] => Aged Brie, 3, 26 
[Aged Brie, 4, 49] => Aged Brie, 3, 50 
[Aged Brie, 4, 50] => Aged Brie, 3, 50 
[Aged Brie, 5, 0] => Aged Brie, 4, 1 
[Aged Brie, 5, 1] => Aged Brie, 4, 2 
[Aged Brie, 5, 25] => Aged Brie, 4, 26 
[Aged Brie, 5, 49] => Aged Brie, 4, 50 
[Aged Brie, 5, 50] => Aged Brie, 4, 50 
[Aged Brie, 6, 0] => Aged Brie, 5, 1 
[Aged Brie, 6, 1] => Aged Brie, 5, 2 
[Aged Brie, 6, 25] => Aged Brie, 5, 26 
[Aged Brie, 6, 49] => Aged Brie, 5, 50 
[Aged Brie, 6, 50] => Aged Brie, 5, 50 
[Aged Brie, 7, 0] => Aged Brie, 6, 1 
[Aged Brie, 7, 1] => Aged Brie, 6, 2 
[Aged Brie, 7, 25] => Aged Brie, 6, 26 
[Aged Brie, 7, 49] => Aged Brie, 6, 50 
[Aged Brie, 7, 50] => Aged Brie, 6, 50 
[Aged Brie, 9, 0] => Aged Brie, 8, 1 
[Aged Brie, 9, 1] => Aged Brie, 8, 2 
[Aged Brie, 9, 25] => Aged Brie, 8, 26 
[Aged Brie, 9, 49] => Aged Brie, 8, 50 
[Aged Brie, 9, 50] => Aged Brie, 8, 50 
[Aged Brie, 10, 0] => Aged Brie, 9, 1 
[Aged Brie, 10, 1] => Aged Brie, 9, 2 
[Aged Brie, 10, 25] => Aged Brie, 9, 26 
[Aged Brie, 10, 49] => Aged Brie, 9, 50 
[Aged Brie, 10, 50] => Aged Brie, 9, 50 
[Aged Brie, 11, 0] => Aged Brie, 10, 1 
[Aged Brie, 11, 1] => Aged Brie, 10, 2 
[Aged Brie, 11, 25] => Aged Brie, 10, 26 
[Aged Brie, 11, 49] => Aged Brie, 10, 50 
[Aged Brie, 11, 50] => Aged Brie, 10, 50 
[Elixir of the Mongoose, -1, 0] => Elixir of the Mongoose, -2, 0 
[Elixir of the Mongoose, -1, 1] => Elixir of the Mongoose, -2, 0 
[Elixir of the Mongoose, -1, 25] => Elixir of the Mongoose, -2, 23 
[Elixir of the Mongoose, -1, 49] => Elixir of the Mongoose, -2, 47 
[Elixir of the Mongoose, -1, 50] => Elixir of the Mongoose, -2, 48 
[Elixir of the Mongoose, 0, 0] => Elixir of the Mongoose, -1, 0 
[Elixir of the Mongoose, 0, 1] => Elixir of the Mongoose, -1, 0 
[Elixir of the Mongoose, 0, 25] => Elixir of the Mongoose, -1, 23 
[Elixir of the Mongoose, 0, 49] => Elixir of the Mongoose, -1, 47 
[Elixir of the Mongoose, 0, 50] => Elixir of the Mongoose, -1, 48 
[Elixir of the Mongoose, 1, 0] => Elixir of the Mongoose, 0, 0 
[Elixir of the Mongoose, 1, 1] => Elixir of the Mongoose, 0, 0 
[Elixir of the Mongoose, 1, 25] => Elixir of the Mongoose, 0, 24 
[Elixir of the Mongoose, 1, 49] => Elixir of the Mongoose, 0, 48 
[Elixir of the Mongoose, 1, 50] => Elixir of the Mongoose, 0, 49 
[Elixir of the Mongoose, 4, 0] => Elixir of the Mongoose, 3, 0 
[Elixir of the Mongoose, 4, 1] => Elixir of the Mongoose, 3, 0 
[Elixir of the Mongoose, 4, 25] => Elixir of the Mongoose, 3, 24 
[Elixir of the Mongoose, 4, 49] => Elixir of the Mongoose, 3, 48 
[Elixir of the Mongoose, 4, 50] => Elixir of the Mongoose, 3, 49 
[Elixir of the Mongoose, 5, 0] => Elixir of the Mongoose, 4, 0 
[Elixir of the Mongoose, 5, 1] => Elixir of the Mongoose, 4, 0 
[Elixir of the Mongoose, 5, 25] => Elixir of the Mongoose, 4, 24 
[Elixir of the Mongoose, 5, 49] => Elixir of the Mongoose, 4, 48 
[Elixir of the Mongoose, 5, 50] => Elixir of the Mongoose, 4, 49 
[Elixir of the Mongoose, 6, 0] => Elixir of the Mongoose, 5, 0 
[Elixir of the Mongoose, 6, 1] => Elixir of the Mongoose, 5, 0 
[Elixir of the Mongoose, 6, 25] => Elixir of the Mongoose, 5, 24 
[Elixir of the Mongoose, 6, 49] => Elixir of the Mongoose, 5, 48 
[Elixir of the Mongoose, 6, 50] => Elixir of the Mongoose, 5, 49 
[Elixir of the Mongoose, 7, 0] => Elixir of the Mongoose, 6, 0 
[Elixir of the Mongoose, 7, 1] => Elixir of the Mongoose, 6, 0 
[Elixir of the Mongoose, 7, 25] => Elixir of the Mongoose, 6, 24 
[Elixir of the Mongoose, 7, 49] => Elixir of the Mongoose, 6, 48 
[Elixir of the Mongoose, 7, 50] => Elixir of the Mongoose, 6, 49 
[Elixir of the Mongoose, 9, 0] => Elixir of the Mongoose, 8, 0 
[Elixir of the Mongoose, 9, 1] => Elixir of the Mongoose, 8, 0 
[Elixir of the Mongoose, 9, 25] => Elixir of the Mongoose, 8, 24 
[Elixir of the Mongoose, 9, 49] => Elixir of the Mongoose, 8, 48 
[Elixir of the Mongoose, 9, 50] => Elixir of the Mongoose, 8, 49 
[Elixir of the Mongoose, 10, 0] => Elixir of the Mongoose, 9, 0 
[Elixir of the Mongoose, 10, 1] => Elixir of the Mongoose, 9, 0 
[Elixir of the Mongoose, 10, 25] => Elixir of the Mongoose, 9, 24 
[Elixir of the Mongoose, 10, 49] => Elixir of the Mongoose, 9, 48 
[Elixir of the Mongoose, 10, 50] => Elixir of the Mongoose, 9, 49 
[Elixir of the Mongoose, 11, 0] => Elixir of the Mongoose, 10, 0 
[Elixir of the Mongoose, 11, 1] => Elixir of the Mongoose, 10, 0 
[Elixir of the Mongoose, 11, 25] => Elixir of the Mongoose, 10, 24 
[Elixir of the Mongoose, 11, 49] => Elixir of the Mongoose, 10, 48 
[Elixir of the Mongoose, 11, 50] => Elixir of the Mongoose, 10, 49 
[Sulfuras, Hand of Ragnaros, -1, 0] => Sulfuras, Hand of Ragnaros, -1, 0 
[Sulfuras, Hand of Ragnaros, -1, 1] => Sulfuras, Hand of Ragnaros, -1, 1 
[Sulfuras, Hand of Ragnaros, -1, 25] => Sulfuras, Hand of Ragnaros, -1, 25 
[Sulfuras, Hand of Ragnaros, -1, 49] => Sulfuras, Hand of Ragnaros, -1, 49 
[Sulfuras, Hand of Ragnaros, -1, 50] => Sulfuras, Hand of Ragnaros, -1, 50 
[Sulfuras, Hand of Ragnaros, 0, 0] => Sulfuras, Hand of Ragnaros, 0, 0 
[Sulfuras, Hand of Ragnaros, 0, 1] => Sulfuras, Hand of Ragnaros, 0, 1 
[Sulfuras, Hand of Ragnaros, 0, 25] => Sulfuras, Hand of Ragnaros, 0, 25 
[Sulfuras, Hand of Ragnaros, 0, 49] => Sulfuras, Hand of Ragnaros, 0, 49 
[Sulfuras, Hand of Ragnaros, 0, 50] => Sulfuras, Hand of Ragnaros, 0, 50 
[Sulfuras, Hand of Ragnaros, 1, 0] => Sulfuras, Hand of Ragnaros, 1, 0 
[Sulfuras, Hand of Ragnaros, 1, 1] => Sulfuras, Hand of Ragnaros, 1, 1 
[Sulfuras, Hand of Ragnaros, 1, 25] => Sulfuras, Hand of Ragnaros, 1, 25 
[Sulfuras, Hand of Ragnaros, 1, 49] => Sulfuras, Hand of Ragnaros, 1, 49 
[Sulfuras, Hand of Ragnaros, 1, 50] => Sulfuras, Hand of Ragnaros, 1, 50 
[Sulfuras, Hand of Ragnaros, 4, 0] => Sulfuras, Hand of Ragnaros, 4, 0 
[Sulfuras, Hand of Ragnaros, 4, 1] => Sulfuras, Hand of Ragnaros, 4, 1 
[Sulfuras, Hand of Ragnaros, 4, 25] => Sulfuras, Hand of Ragnaros, 4, 25 
[Sulfuras, Hand of Ragnaros, 4, 49] => Sulfuras, Hand of Ragnaros, 4, 49 
[Sulfuras, Hand of Ragnaros, 4, 50] => Sulfuras, Hand of Ragnaros, 4, 50 
[Sulfuras, Hand of Ragnaros, 5, 0] => Sulfuras, Hand of Ragnaros, 5, 0 
[Sulfuras, Hand of Ragnaros, 5, 1] => Sulfuras, Hand of Ragnaros, 5, 1 
[Sulfuras, Hand of Ragnaros, 5, 25] => Sulfuras, Hand of Ragnaros, 5, 25 
[Sulfuras, Hand of Ragnaros, 5, 49] => Sulfuras, Hand of Ragnaros, 5, 49 
[Sulfuras, Hand of Ragnaros, 5, 50] => Sulfuras, Hand of Ragnaros, 5, 50 
[Sulfuras, Hand of Ragnaros, 6, 0] => Sulfuras, Hand of Ragnaros, 6, 0 
[Sulfuras, Hand of Ragnaros, 6, 1] => Sulfuras, Hand of Ragnaros, 6, 1 
[Sulfuras, Hand of Ragnaros, 6, 25] => Sulfuras, Hand of Ragnaros, 6, 25 
[Sulfuras, Hand of Ragnaros, 6, 49] => Sulfuras, Hand of Ragnaros, 6, 49 
[Sulfuras, Hand of Ragnaros, 6, 50] => Sulfuras, Hand of Ragnaros, 6, 50 
[Sulfuras, Hand of Ragnaros, 7, 0] => Sulfuras, Hand of Ragnaros, 7, 0 
[Sulfuras, Hand of Ragnaros, 7, 1] => Sulfuras, Hand of Ragnaros, 7, 1 
[Sulfuras, Hand of Ragnaros, 7, 25] => Sulfuras, Hand of Ragnaros, 7, 25 
[Sulfuras, Hand of Ragnaros, 7, 49] => Sulfuras, Hand of Ragnaros, 7, 49 
[Sulfuras, Hand of Ragnaros, 7, 50] => Sulfuras, Hand of Ragnaros, 7, 50 
[Sulfuras, Hand of Ragnaros, 9, 0] => Sulfuras, Hand of Ragnaros, 9, 0 
[Sulfuras, Hand of Ragnaros, 9, 1] => Sulfuras, Hand of Ragnaros, 9, 1 
[Sulfuras, Hand of Ragnaros, 9, 25] => Sulfuras, Hand of Ragnaros, 9, 25 
[Sulfuras, Hand of Ragnaros, 9, 49] => Sulfuras, Hand of Ragnaros, 9, 49 
[Sulfuras, Hand of Ragnaros, 9, 50] => Sulfuras, Hand of Ragnaros, 9, 50 
[Sulfuras, Hand of Ragnaros, 10, 0] => Sulfuras, Hand of Ragnaros, 10, 0 
[Sulfuras, Hand of Ragnaros, 10, 1] => Sulfuras, Hand of Ragnaros, 10, 1 
[Sulfuras, Hand of Ragnaros, 10, 25] => Sulfuras, Hand of Ragnaros, 10, 25 
[Sulfuras, Hand of Ragnaros, 10, 49] => Sulfuras, Hand of Ragnaros, 10, 49 
[Sulfuras, Hand of Ragnaros, 10, 50] => Sulfuras, Hand of Ragnaros, 10, 50 
[Sulfuras, Hand of Ragnaros, 11, 0] => Sulfuras, Hand of Ragnaros, 11, 0 
[Sulfuras, Hand of Ragnaros, 11, 1] => Sulfuras, Hand of Ragnaros, 11, 1 
[Sulfuras, Hand of Ragnaros, 11, 25] => Sulfuras, Hand of Ragnaros, 11, 25 
[Sulfuras, Hand of Ragnaros, 11, 49] => Sulfuras, Hand of Ragnaros, 11, 49 
[Sulfuras, Hand of Ragnaros, 11, 50] => Sulfuras, Hand of Ragnaros, 11, 50 
[Backstage passes to a TAFKAL80ETC concert, -1, 0] => Backstage passes to a TAFKAL80ETC concert, -2, 0 
[Backstage passes to a TAFKAL80ETC concert, -1, 1] => Backstage passes to a TAFKAL80ETC concert, -2, 0 
[Backstage passes to a TAFKAL80ETC concert, -1, 25] => Backstage passes to a TAFKAL80ETC concert, -2, 0 
[Backstage passes to a TAFKAL80ETC concert, -1, 49] => Backstage passes to a TAFKAL80ETC concert, -2, 0 
[Backstage passes to a TAFKAL80ETC concert, -1, 50] => Backstage passes to a TAFKAL80ETC concert, -2, 0 
[Backstage passes to a TAFKAL80ETC concert, 0, 0] => Backstage passes to a TAFKAL80ETC concert, -1, 0 
[Backstage passes to a TAFKAL80ETC concert, 0, 1] => Backstage passes to a TAFKAL80ETC concert, -1, 0 
[Backstage passes to a TAFKAL80ETC concert, 0, 25] => Backstage passes to a TAFKAL80ETC concert, -1, 0 
[Backstage passes to a TAFKAL80ETC concert, 0, 49] => Backstage passes to a TAFKAL80ETC concert, -1, 0 
[Backstage passes to a TAFKAL80ETC concert, 0, 50] => Backstage passes to a TAFKAL80ETC concert, -1, 0 
[Backstage passes to a TAFKAL80ETC concert, 1, 0] => Backstage passes to a TAFKAL80ETC concert, 0, 3 
[Backstage passes to a TAFKAL80ETC concert, 1, 1] => Backstage passes to a TAFKAL80ETC concert, 0, 4 
[Backstage passes to a TAFKAL80ETC concert, 1, 25] => Backstage passes to a TAFKAL80ETC concert, 0, 28 
[Backstage passes to a TAFKAL80ETC concert, 1, 49] => Backstage passes to a TAFKAL80ETC concert, 0, 50 
[Backstage passes to a TAFKAL80ETC concert, 1, 50] => Backstage passes to a TAFKAL80ETC concert, 0, 50 
[Backstage passes to a TAFKAL80ETC concert, 4, 0] => Backstage passes to a TAFKAL80ETC concert, 3, 3 
[Backstage passes to a TAFKAL80ETC concert, 4, 1] => Backstage passes to a TAFKAL80ETC concert, 3, 4 
[Backstage passes to a TAFKAL80ETC concert, 4, 25] => Backstage passes to a TAFKAL80ETC concert, 3, 28 
[Backstage passes to a TAFKAL80ETC concert, 4, 49] => Backstage passes to a TAFKAL80ETC concert, 3, 50 
[Backstage passes to a TAFKAL80ETC concert, 4, 50] => Backstage passes to a TAFKAL80ETC concert, 3, 50 
[Backstage passes to a TAFKAL80ETC concert, 5, 0] => Backstage passes to a TAFKAL80ETC concert, 4, 3 
[Backstage passes to a TAFKAL80ETC concert, 5, 1] => Backstage passes to a TAFKAL80ETC concert, 4, 4 
[Backstage passes to a TAFKAL80ETC concert, 5, 25] => Backstage passes to a TAFKAL80ETC concert, 4, 28 
[Backstage passes to a TAFKAL80ETC concert, 5, 49] => Backstage passes to a TAFKAL80ETC concert, 4, 50 
[Backstage passes to a TAFKAL80ETC concert, 5, 50] => Backstage passes to a TAFKAL80ETC concert, 4, 50 
[Backstage passes to a TAFKAL80ETC concert, 6, 0] => Backstage passes to a TAFKAL80ETC concert, 5, 2 
[Backstage passes to a TAFKAL80ETC concert, 6, 1] => Backstage passes to a TAFKAL80ETC concert, 5, 3 
[Backstage passes to a TAFKAL80ETC concert, 6, 25] => Backstage passes to a TAFKAL80ETC concert, 5, 27 
[Backstage passes to a TAFKAL80ETC concert, 6, 49] => Backstage passes to a TAFKAL80ETC concert, 5, 50 
[Backstage passes to a TAFKAL80ETC concert, 6, 50] => Backstage passes to a TAFKAL80ETC concert, 5, 50 
[Backstage passes to a TAFKAL80ETC concert, 7, 0] => Backstage passes to a TAFKAL80ETC concert, 6, 2 
[Backstage passes to a TAFKAL80ETC concert, 7, 1] => Backstage passes to a TAFKAL80ETC concert, 6, 3 
[Backstage passes to a TAFKAL80ETC concert, 7, 25] => Backstage passes to a TAFKAL80ETC concert, 6, 27 
[Backstage passes to a TAFKAL80ETC concert, 7, 49] => Backstage passes to a TAFKAL80ETC concert, 6, 50 
[Backstage passes to a TAFKAL80ETC concert, 7, 50] => Backstage passes to a TAFKAL80ETC concert, 6, 50 
[Backstage passes to a TAFKAL80ETC concert, 9, 0] => Backstage passes to a TAFKAL80ETC concert, 8, 2 
[Backstage passes to a TAFKAL80ETC concert, 9, 1] => Backstage passes to a TAFKAL80ETC concert, 8, 3 
[Backstage passes to a TAFKAL80ETC concert, 9, 25] => Backstage passes to a TAFKAL80ETC concert, 8, 27 
[Backstage passes to a TAFKAL80ETC concert, 9, 49] => Backstage passes to a TAFKAL80ETC concert, 8, 50 
[Backstage passes to a TAFKAL80ETC concert, 9, 50] => Backstage passes to a TAFKAL80ETC concert, 8, 50 
[Backstage passes to a TAFKAL80ETC concert, 10, 0] => Backstage passes to a TAFKAL80ETC concert, 9, 2 
[Backstage passes to a TAFKAL80ETC concert, 10, 1] => Backstage passes to a TAFKAL80ETC concert, 9, 3 
[Backstage passes to a TAFKAL80ETC concert, 10, 25] => Backstage passes to a TAFKAL80ETC concert, 9, 27 
[Backstage passes to a TAFKAL80ETC concert, 10, 49] => Backstage passes to a TAFKAL80ETC concert, 9, 50 
[Backstage passes to a TAFKAL80ETC concert, 10, 50] => Backstage passes to a TAFKAL80ETC concert, 9, 50 
[Backstage passes to a TAFKAL80ETC concert, 11, 0] => Backstage passes to a TAFKAL80ETC concert, 10, 1 
[Backstage passes to a TAFKAL80ETC concert, 11, 1] => Backstage passes to a TAFKAL80ETC concert, 10, 2 
[Backstage passes to a TAFKAL80ETC concert, 11, 25] => Backstage passes to a TAFKAL80ETC concert, 10, 26 
[Backstage passes to a TAFKAL80ETC concert, 11, 49] => Backstage passes to a TAFKAL80ETC concert, 10, 50 
[Backstage passes to a TAFKAL80ETC concert, 11, 50] => Backstage passes to a TAFKAL80ETC concert, 10, 50 
[Conjured item, -1, 0] => Conjured item, -2, 0 
[Conjured item, -1, 1] => Conjured item, -2, 0 
[Conjured item, -1, 25] => Conjured item, -2, 21 
[Conjured item, -1, 49] => Conjured item, -2, 45 
[Conjured item, -1, 50] => Conjured item, -2, 46 
[Conjured item, 0, 0] => Conjured item, -1, 0 
[Conjured item, 0, 1] => Conjured item, -1, 0 
[Conjured item, 0, 25] => Conjured item, -1, 21 
[Conjured item, 0, 49] => Conjured item, -1, 45 
[Conjured item, 0, 50] => Conjured item, -1, 46 
[Conjured item, 1, 0] => Conjured item, 0, 0 
[Conjured item, 1, 1] => Conjured item, 0, 0 
[Conjured item, 1, 25] => Conjured item, 0, 23 
[Conjured item, 1, 49] => Conjured item, 0, 47 
[Conjured item, 1, 50] => Conjured item, 0, 48 
[Conjured item, 4, 0] => Conjured item, 3, 0 
[Conjured item, 4, 1] => Conjured item, 3, 0 
[Conjured item, 4, 25] => Conjured item, 3, 23 
[Conjured item, 4, 49] => Conjured item, 3, 47 
[Conjured item, 4, 50] => Conjured item, 3, 48 
[Conjured item, 5, 0] => Conjured item, 4, 0 
[Conjured item, 5, 1] => Conjured item, 4, 0 
[Conjured item, 5, 25] => Conjured item, 4, 23 
[Conjured item, 5, 49] => Conjured item, 4, 47 
[Conjured item, 5, 50] => Conjured item, 4, 48 
[Conjured item, 6, 0] => Conjured item, 5, 0 
[Conjured item, 6, 1] => Conjured item, 5, 0 
[Conjured item, 6, 25] => Conjured item, 5, 23 
[Conjured item, 6, 49] => Conjured item, 5, 47 
[Conjured item, 6, 50] => Conjured item, 5, 48 
[Conjured item, 7, 0] => Conjured item, 6, 0 
[Conjured item, 7, 1] => Conjured item, 6, 0 
[Conjured item, 7, 25] => Conjured item, 6, 23 
[Conjured item, 7, 49] => Conjured item, 6, 47 
[Conjured item, 7, 50] => Conjured item, 6, 48 
[Conjured item, 9, 0] => Conjured item, 8, 0 
[Conjured item, 9, 1] => Conjured item, 8, 0 
[Conjured item, 9, 25] => Conjured item, 8, 23 
[Conjured item, 9, 49] => Conjured item, 8, 47 
[Conjured item, 9, 50] => Conjured item, 8, 48 
[Conjured item, 10, 0] => Conjured item, 9, 0 
[Conjured item, 10, 1] => Conjured item, 9, 0 
[Conjured item, 10, 25] => Conjured item, 9, 23 
[Conjured item, 10, 49] => Conjured item, 9, 47 
[Conjured item, 10, 50] => Conjured item, 9, 48 
[Conjured item, 11, 0] => Conjured item, 10, 0 
[Conjured item, 11, 1] => Conjured item, 10, 0 
[Conjured item, 11, 25] => Conjured item, 10, 23 
[Conjured item, 11, 49] => Conjured item, 10, 47 
[Conjured item, 11, 50] => Conjured item, 10, 48 
